### DISCUSSION TITLE
Restrict objects to a defined number in a certain area

### DISCUSSION THREAD

**Discussion by bweinlich** [systemed/tilemaker#500] [Q&A] [answered]
I want to show OSM trees on my map, so I added something like this in the LUA-Script:

	-- Write 'tree'
	if natural == "tree" then
		node:Layer("tree", false)
		node:Attribute("natural", "tree")
		node:Attribute("leaf_type", node:Find("leaf_type"))
		return
	end

This works very well but I recognized that in some cities there are really lots of trees which is a bit overwhelming in certain areas. So I wonder if it is possible to restrict the density of trees (or any other objects). For example a maximum of 10 trees in a 10x10 metres square.

So I would need something like a spatial query around the current object. Is this possible in the Lua script or is there something like a postprocessing step that I can hook into?

Thanks in advance :)

**Comment by systemed** [comment, marked as answer]
This sort of heavy geometry processing isn't really something that tilemaker does, I'm afraid - you'd typically use PostGIS for this. [Martin](https://martin.maplibre.org) would be really well suited for serving vector tiles with queries like this.

If you did want to still use tilemaker, you'd typically load the trees into a Postgres database (using osm2pgsql and some custom Lua); cluster them with PostGIS functions; then use pgsql2shp to export the clustered trees to a shapefile which tilemaker could load.

**Reply by bweinlich** [reply]
I'm afraid that computing something like this on the fly with Martin could cost a lot of computing time. As my tiles are hosted inside a public cloud I would like reduce computing time to a minimum.

But your second idea using osm2pgsql in combination with PostGIS & QGIS as frontend should work well as QGIS has a point clusterin feature. Thank you very much!

**Reply by bweinlich** [reply]
I tried it out and now I have a Shapefile with the trees as point features. I added the layer in the config.json like this:

"tree":              { "minzoom": 13, "maxzoom": 14, "source": "/tiles/trees-reduced-shape/trees-reduced.shp", "source_columns": true }

Tile creation works without errors (the shapefile is listed). But now I can't see the trees in the resulting vector tiles anymore. Do I have to reference the layer in the process.lua file? Have I forgotten something?

**Reply by systemed** [reply]
Is your shapefile in standard lat/long projection (EPSG 4326)?

**Reply by bweinlich** [reply]
It was in WebMercator. Thankyou, this did the trick. Sry for not looking in the documentation.

**Reply by systemed** [reply]
No problem - glad it works!

**Comment by dieterdreist** [comment]
mapbox/tippecanoe can reduce the number of point features according to a density (max number in tile)

**Reply by bweinlich** [reply]
Can tippecanoe also process osm.pbf files as input like tilemaker? Or do I need a preprocessing here?

I think I will use PostGIS & QGIS because I am more flexible with it. Nevertheless many thanks for your idea :)

**Comment by dieterdreist** [comment]
sent from a phone

> On 18 Jun 2023, at 13:46, ToraXaroT ***@***.***> wrote:
> 
> Can tippecanoe also process osm.pbf files as input like tilemaker? Or do I need a preprocessing here?


afaik only geojson as input.
